predicators/nsrt_learning/strips_learning/effect_search_learner.py

```diff
@@ -329,77 +327,64 @@ def get_successors(
                                              get_successors=get_successors,
                                              heuristic=heuristic)
 
-        # Extract the best effect set.
-        best_effect_sets = path[-1]
-        # Convert into PNADs.
-        final_pnads = self._effect_sets_to_pnads(best_effect_sets)
-        return final_pnads
+        # Extract the best PNADs set.
+        final_pnads = path[-1]
+        # Recompute these PNADs so that they exactly match the PNADs used
+        # to compute the final heuristic.
+        recomp_final_pnads = self._recompute_pnads_from_effects(
+            list(final_pnads))
+        # Fix naming.
+        pnad_map: Dict[ParameterizedOption, List[PartialNSRTAndDatastore]] = {
+            p.option_spec[0]: []
+            for p in recomp_final_pnads
+        }
+        for p in recomp_final_pnads:
+            p.op = p.op.copy_with(name=p.option_spec[0].name)
+            pnad_map[p.option_spec[0]].append(p)
+        ret_pnads = self._get_uniquely_named_nec_pnads(pnad_map)
+        return ret_pnads
+
+    @functools.lru_cache(maxsize=None)
+    def _create_general_pnad_for_option(
+            self, parameterized_option: ParameterizedOption
+    ) -> PartialNSRTAndDatastore:
+        """Create the most general PNAD for the given option."""
+        # Create the parameters, which are determined solely from the option
+        # types, since the most general operator has no add/delete effects.
+        parameters = utils.create_new_variables(parameterized_option.types)
+        option_spec = (parameterized_option, parameters)
+        # In the most general operator, the ignore effects contain ALL
+        # predicates.
+        ignore_effects = self._predicates.copy()
+        # There are no add effects or delete effects. The preconditions
+        # are initialized to be trivial. They will be recomputed next.
+        op = STRIPSOperator(parameterized_option.name, parameters, set(),
+                            set(), set(), ignore_effects)
+        pnad = PartialNSRTAndDatastore(op, [], option_spec)
+        return pnad
```

Comment:
Note that this method is basically the same as the superclass one except that it doesn't induce operator preconditions (doing so would trigger an `AssertionError` for this learner since the PNAD won't have any data assigned to it!). Maybe we should modify the superclass to not compute preconditions (that shouldn't affect the `BackchainingSTRIPSLearner`, but it might...)? I couldn't think of a nice SWE way to minimize code duplication here.

Reply:
wow TIL you can comment on multiple lines in code reviews!!

Reply:
>Maybe we should modify the superclass to not compute preconditions (that shouldn't affect the BackchainingSTRIPSLearner, but it might...)?

yeah let's try this